lib/sycamore/sycamore/document.py

```diff
@@ -0,0 +1,388 @@
+from abc import abstractmethod, ABC
```

Comment:
Is your intention to merge this PR into Sycamore, or just to use it as a vehicle to discuss the data model? 

(To be clear, I want to get this into Sycamore, just want to get a sense of whether we should discuss the nuts and bolts now defer that).